### DISCUSSION TITLE
DietPi-Software | Podman

### DISCUSSION THREAD

**Discussion by isarrider** [MichaIng/DietPi#5344] [Software Request]
has anybody looked in podman on DietPi so far?
it is on my plan to look into it...

**Comment by Joulinar** [comment]
not that I'm aware of.

**Comment by MichaIng** [comment]
Adding the GitHub link: https://github.com/containers/podman

**Comment by queeup** [comment]
Personally I prefer podman.

**Comment by isarrider** [comment]
from many POVs, podman is superior...
If I find time, Ill investigate a bit more
and post my findings here...

**Reply by Joulinar** [reply]
following is not working??

```
apt install podman
```

**Reply by radekbenkel** [reply]
I was able to have running `podman` on DietPi. 

Instructions:

* install `apt install podman uidmap slirp4netns`

After that you will get:

```shell
dietpi@DietPi:~$ podman ps
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available
WARN[0000] For using systemd, you may need to login using an user session
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1000` (possibly as root)
WARN[0000] Falling back to --cgroup-manager=cgroupfs
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available
WARN[0000] For using systemd, you may need to login using an user session
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1000` (possibly as root)
WARN[0000] Falling back to --cgroup-manager=cgroupfs
```

* to suppress that warning, create directory `mkdir -p ~/.config/containers` with following file: 

```shell
dietpi@DietPi:~$ cat ~/.config/containers/containers.conf
[engine]
events_logger = "file"
cgroup_manager = "cgroupfs"
```

* after these two steps there's no errors anymore: 

```shell
dietpi@DietPi:~$ podman ps
CONTAINER ID  IMAGE   COMMAND  CREATED  STATUS  PORTS   NAMES
```

* **optional:** one might want to add `docker.io` as `unqualified-search-registry`, so you can use for example `alpine` instead of `docker.io/alpine`. 

```shell
dietpi@DietPi:~$ cat ~/.config/containers/registries.conf
unqualified-search-registries = ['docker.io'] // or quay.io or something else 
```

* and then:

```shell
dietpi@DietPi:~$ podman run -it --rm alpine ash
# [... will download an image if you don't have it yet] 
/ # echo "hello from alpine running in a rootless container via podman!"
hello from alpine running in a rootless container via podman!
```

**Reply by Joulinar** [reply]
what system you are running? Because on my R5S, it just requires `podman` as well as `dbus` to be installed only. No need to manage the `cgroupv2` error. I guess it's related to kernel version.

**Reply by radekbenkel** [reply]
@Joulinar sorry, forgot to mention it above :)

```
 ─────────────────────────────────────────────────────
 DietPi v8.10.2 : 18:39 - Mon 10/24/22
 ─────────────────────────────────────────────────────
 - Device model : RPi 3 Model B (aarch64)
 - CPU temp : 36 °C / 96 °F : Cool runnings
 ─────────────────────────────────────────────────────
 
 $ uname -a 
 Linux DietPi 5.15.61-v8+ #1579 SMP PREEMPT Fri Aug 26 11:16:44 BST 2022 aarch64 GNU/Linux
```

EDIT: tried on VirtualBox's VM and got similar errors, `cgroupv2` and `newuidmap` missing. 

Maybe in your case other packages added what's required and you can use `cgroupManager: systemd` instead of `cgroupManager: cgroupfs` like in my case (based on `podman info` output) 🤔

**Reply by Joulinar** [reply]
ok I tested on the same device and it's working ootb

```
apt install podman dbus
systemctl enable dbus
systemctl start dbus
podman run -dt -p 8080:80/tcp docker.io/library/httpd
```

and `podman ps` returns following

```
root@DietPi:~# podman ps
CONTAINER ID  IMAGE                           COMMAND           CREATED             STATUS                 PORTS                 NAMES
dc634e1cc46d  docker.io/library/httpd:latest  httpd-foreground  About a minute ago  Up About a minute ago  0.0.0.0:8080->80/tcp  jovial_edison
root@DietPi:~#
```

**Reply by radekbenkel** [reply]
@Joulinar ok, now it hit me with idea why we have different behavior. 

From your examples: `root@DietPi:~# podman ps` ... `root@DietPi` ... `root`. I replicated your instructions as a root (on x86 VM) and indeed everything worked (with `cgroupManager` ended up being `systemd`). 

My instructions were from a POV of non-root user, as one of the selling points of `podman` is ability to run *rootless* containers. My assumption is that's why I needed `uidmap` (for UID mapping) and `slirp4netns` (from package description: "User-mode networking for unprivileged network namespaces").

`htop` output for both cases seem to confirm my assumptions: 

* as `dietpi`:

![dietpi](https://user-images.githubusercontent.com/218345/197786367-c401c50f-9f58-48c8-bcfb-3501f7b52015.png)

* as `root`:

![root](https://user-images.githubusercontent.com/218345/197786376-224efe14-5641-4be4-95b7-f1aa4c3737cc.png)

Does it make sense? (again, my Linux-fu is far from being top-notch :)

**Reply by Joulinar** [reply]
most probably you are right on your assumption with the rootless user setup https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md

**Comment by Gibbz** [comment]
Im trying to setup podman on dietpi as rootless user. I'm having issues with environment variables.
I get the errors when using systemd in user mode:
```
systemctl --user daemon-reexec && systemctl --user daemon-reload

Failed to connect to user scope bus via local transport: $DBUS_SESSION_BUS_ADDRESS and $XDG_RUNTIME_DIR not defined (consider using --machine=<user>@.host --user to connect to bus of other user)
```

Ive tried setting the variables in, but they always echo as blank....
```
/etc/environment
$HOME/.config/environment.d/60-bus-fix.conf 
```

**Reply by Gibbz** [reply]
Looks like it was a dropbear issue, changed to openssh!

**Reply by MichaIng** [reply]
Right, Dropbear has only limited PAM support, hence no automatic dbus/logind registration for user services.

**Comment by Gibbz** [comment]
So I've setup podman with a bunch of containers. 
I'm having a real issue with jellyfin. The container keeps shutting down during library scans. Then it takes some time to come back up. The logs indicate systemd is shutting it down.....

Is dietpi doing some over optimisations with systemd to cause it to keep shutting down my podman services?

**Reply by Joulinar** [reply]
DietPi is doing nothing to any application during runtime.

**Reply by MichaIng** [reply]
Probably OOM killed? Check `htop` for memory usage while scanning the library, and kernel logs:
```sh
dmesg -l 0,1,2,3
```

**Reply by Gibbz** [reply]
Thanks... dmesg doesnt show anything and the system has 16g of ram (no swap though).
I'll run it without systemd and see if that makes any difference.

**Reply by Gibbz** [reply]
Ok not sure what's going on. If I run using podman -d it works fine. But when using systemd it keeps getting terminated. 

I'm pretty sure I've enabled linger, and it's not out of memory... 

A strange thing I've noticed. If I login to cockpit and check the pod status it always says created less than a minute ago, and running less than a minute. Maybe the linger is not working. And the services are killed...

Edit: looks like linger didn't work. I've reenabled it and will try again!

**Reply by MichaIng** [reply]
Does it work when you start them as system services as root?

`systemd-logind` is unmasked/running?

**Reply by Gibbz** [reply]
Reenabling linger mode looks to have fixed the issue 👍 no problem since. 
I don't know why the command didn't stick the first time when I set it up.

**Reply by MichaIng** [reply]
Okay so it was a user service thing then. To be honest, I have near zero experience with those, as I personally set up everything as system service with dedicated service user.

Just know that it generally requires logind registration, PAM, anf hence DBus for unprivileged users to spawn and access a user systemd instance, and that linger is needed to keep those user instances active after the session ended.
